content/blog/tidyselect-1-2-0/index.Rmd

```diff
@@ -0,0 +1,158 @@
+mtcars |> select(all_of(letters))
+```
+
+The latter is more lenient and ignores any names that are not present in the data frame. In this case, it ends up selecting nothing:
+
+```{r}
+mtcars |> select(any_of(letters))
+```
+
+Another feature of `all_of()` and `any_of()` is that they remove all ambiguity between variables in your environment like `vars` or `letters` (env-variables) and variables inside the data frame like `cyl` or `am` (data-variables). Let's add `vars` in the data frame to see what happens:
+
+```{r}
+my_data <- mtcars |> mutate(vars = 1:n())
+my_data |> select(all_of(vars)) |> glimpse()
+```
+
+Because `vars` was supplied to `all_of()`, `select()` will never confuse it with `mtcars$vars`. In technical terms, there is no __data-masking__ within selection helpers like `all_of()`, `any_of()`, or even `starts_with()`. It is safe to supply env-variables to these functions without worrying about data-masking ambiguity.
+
+This is not the case however if you supply a character vector outside of `all_of()`:
+
+```{r}
+my_data |> select(vars) |> glimpse()
+```
+
+This is why we have decided to deprecate direct supply of character vectors in favour of using `all_of()` and `any_of()`. You will now get a soft-deprecation warning recommending to use `all_of()`:
+
+```{r}
+mtcars |> select(vars) |> glimpse()
+```
+
+
+## Using `.data` inside selections
+
+The `.data` pronoun is a convenient way of programming with data-masking functions like `mutate()` and `filter()`. It has two main functions:
+
+1. Retrieve a data frame column from a name stored in a variable with `[[`.
+
+   ```{r}
+   var <- "am"
+   mtcars |> transmute(am = .data[[var]] * 10) |> glimpse()
+   ```
+
+2. For package developers, `.data` is helpful to silence R CMD check notes about unknown variables. When the static analysis checker of R encounters an expression like `mtcars |> mutate(am * 2)`, it has no way of knowing that `am` is a data frame variable. Since it doesn't see any variable `am` in your environment, it emits a warning about a potential typo in the code.
+
+   The `.data$col` pattern is used to work around this issue: `mtcars |> mutate(.data$am * 2)` doesn't produce any warnings.
+
+Whereas `.data` is very useful in data-masking functions, its usage in selections is much more limited. As we have seen in the previous section, retrieving a variable from character vector should be done with `all_of()`:
+
+```{r}
+var <- "am"
+mtcars |> select(all_of(var)) |> glimpse()
+```
+
+And to avoid the R CMD check note about unknown variables, it is much cleaner to wrap the column name in quotes:
+
+```{r}
+mtcars |> select("am") |> glimpse()
+```
+
+Allowing the `.data` pronoun in selection contexts also makes the distinction between tidy-selections and data-masking blurrier. And so we have decided to deprecate it in selections:
+
+```{r}
+var <- "am"
+mtcars |> select(.data[[var]]) |> invisible()
+```
+
+```{r}
+mtcars |> select(.data$am) |> invisible()
+```
+
```

Comment:
```suggestion
This deprecation does not affect the use of `.data` in data-masking contexts.
```

Reply:
I realize this is redundant, but I'm just being extra explicit, since this was the source of confusion.